File: packages/conformance/src/consumer.ts
import type { AssuranceDecision } from '@l9/assurance-contracts';
import { canonicalJson, sha256Digest } from '@l9/assurance-evidence';
import { escapeMarkdown, renderDecisionSummary, verifyDecision } from '@l9/assurance-evaluator';

export interface ConsumerConformanceReport {
  readonly consumerId: string;
  readonly passed: boolean;
  readonly checks: readonly { readonly id: string; readonly passed: boolean; readonly message: string }[];
}

export function runConsumerConformance(input: {
  readonly consumerId: string;
  readonly canonicalDecision: unknown;
  readonly transportedDecisionText: string;
  readonly publishedVerdict: string;
  readonly publishedSummary: string;
}): ConsumerConformanceReport {
  const verification = verifyDecision(input.canonicalDecision);
  let canonical = '';
  let transported: unknown;
  let canonicalDigest = '';
  let transportedDigest = '';
  let expectedSummary = '';
  let expectedVerdict = '';

  try {
    canonical = `${canonicalJson(input.canonicalDecision)}\n`;
    canonicalDigest = sha256Digest(input.canonicalDecision).value;
  } catch {
    canonical = '';
  }
  try {
    transported = JSON.parse(input.transportedDecisionText) as unknown;
    transportedDigest = sha256Digest(transported).value;
  } catch {
    transported = undefined;
  }
  if (verification.valid && isRecord(input.canonicalDecision)) {
    expectedVerdict = typeof input.canonicalDecision.verdict === 'string' ? input.canonicalDecision.verdict : '';
    try {
      expectedSummary = renderDecisionSummary(input.canonicalDecision as unknown as AssuranceDecision);
    } catch {
      expectedSummary = '';
    }
  }

  const checks = [
    check('byte-transport', canonical.length > 0 && input.transportedDecisionText === canonical, 'decision transported byte-for-byte'),
    check('digest', canonicalDigest.length > 0 && transportedDigest === canonicalDigest, 'decision digest preserved'),
    check('verdict', expectedVerdict.length > 0 && input.publishedVerdict.toLowerCase() === expectedVerdict, 'published verdict matches'),
    check('summary', expectedSummary.length > 0 && input.publishedSummary === expectedSummary, 'summary is canonical projection'),
    check('schema', verification.valid, 'decision schema supported'),
    check('escaping', !input.publishedSummary.includes('<script>'), 'summary escapes raw HTML'),
  ];
  return {
    consumerId: input.consumerId,
    passed: checks.every((item) => item.passed),
    checks: Object.freeze(checks),
  };
}

export function escapeConsumerText(value: string): string {
  return escapeMarkdown(value);
}

function check(
  id: string,
  passed: boolean,
  message: string,
): { readonly id: string; readonly passed: boolean; readonly message: string } {
  return { id, passed, message };
}

function isRecord(value: unknown): value is Readonly<Record<string, unknown>> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: src/l9_assurance/conformance/consumer.py
from __future__ import annotations

import json
from typing import Any, Mapping

from l9_assurance.evaluator import escape_markdown, render_decision_summary, verify_decision
from l9_assurance.evidence import canonical_json, sha256_digest


def run_consumer_conformance(
    *,
    consumer_id: str,
    canonical_decision: Any,
    transported_decision_text: str,
    published_verdict: str,
    published_summary: str,
) -> dict[str, Any]:
    verification = verify_decision(canonical_decision)
    try:
        canonical = canonical_json(canonical_decision) + "\n"
        canonical_digest = sha256_digest(canonical_decision)["value"]
    except (TypeError, ValueError):
        canonical = ""
        canonical_digest = ""
    try:
        transported = json.loads(transported_decision_text)
        transported_digest = sha256_digest(transported)["value"]
    except (json.JSONDecodeError, TypeError, ValueError):
        transported_digest = ""
    expected_verdict = canonical_decision.get("verdict", "") if isinstance(canonical_decision, Mapping) and verification["valid"] else ""
    try:
        expected_summary = render_decision_summary(canonical_decision) if verification["valid"] else ""
    except (KeyError, TypeError):
        expected_summary = ""
    checks = [
        _check("byte-transport", bool(canonical) and transported_decision_text == canonical, "decision transported byte-for-byte"),
        _check("digest", bool(canonical_digest) and transported_digest == canonical_digest, "decision digest preserved"),
        _check("verdict", bool(expected_verdict) and published_verdict.lower() == expected_verdict, "published verdict matches"),
        _check("summary", bool(expected_summary) and published_summary == expected_summary, "summary is canonical projection"),
        _check("schema", verification["valid"], "decision schema supported"),
        _check("escaping", "<script>" not in published_summary, "summary escapes raw HTML"),
    ]
    return {"consumerId": consumer_id, "passed": all(item["passed"] for item in checks), "checks": checks}


def escape_consumer_text(value: str) -> str:
    return escape_markdown(value)


def _check(identity: str, passed: bool, message: str) -> dict[str, Any]:
    return {"id": identity, "passed": passed, "message": message}
